### PR TITLE
Fix race-condition when Driver re-consults Operator.isBlocked

### PR DESCRIPTION
**Pull request by findepi**
Fixes second problem described in https://github.com/prestodb/presto/issues/10003:

Previously, if `Driver` couldn't make progress (`!movedPage`), we listed
all blocked operators and waited until any of these unblocked. However,
IF an operator become unblocked between the time when it was considered by
the processing loop AND the time when we construct driver's composite
is-blocked future THEN the this fact is lost. Thus, even though an
operator became unblocked, driver will still be blocked, waiting for
something else.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/Driver.java [hunk 1]
```diff
@@ -363,32 +365,42 @@ private void processNewSources()
...
+                if (!current.isFinished()) {
+                    Optional<ListenableFuture<?>> nextBlocked = getBlockedFuture(next);
+                    if (nextBlocked.isPresent()) {
+                        addBlockedOperator(blockedOperators, blockedFutures, next, nextBlocked.get());
                     }
-
-                    if (current instanceof SourceOperator) {
-                        movedPage = true;
+                    else if (next.needsInput()) {
+                        // get an output page from current operator
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: Dear reviewer,

the code block wasn't change. Just the indentation.

File: presto-main/src/main/java/com/facebook/presto/operator/Driver.java [hunk 2]
```diff
@@ -402,29 +414,18 @@ private void processNewSources()
...
-                for (Operator operator : operators) {
-                    Optional<ListenableFuture<?>> blocked = getBlockedFuture(operator);
-                    if (blocked.isPresent()) {
-                        blockedOperators.add(operator);
-                        blockedFutures.add(blocked.get());
-                    }
-                }
-
-                if (!blockedFutures.isEmpty()) {
-                    // unblock when the first future is complete
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: Dear reviewer,

the code block wasn't change. Just the indentation.

File: presto-main/src/main/java/com/facebook/presto/operator/Driver.java [hunk 3]
```diff
@@ -447,6 +448,28 @@ private void processNewSources()
...
+            // We don't want to duplicate operators on the list. Just update its "entry" with new blocked future
+            blockedFutures.set(blockedFutures.size() - 1, blockedFuture);
+            return;
+        }
+
+        blockedOperators.add(operator);
+        blockedFutures.add(blockedFuture);
+    }
+
+    private static boolean allBlocked(List<ListenableFuture<?>> blockedFutures)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: I think we can live without this optimization. The code is long and complicated without it enough.

File: presto-main/src/main/java/com/facebook/presto/operator/Driver.java [hunk 4]
```diff
@@ -447,6 +448,28 @@ private void processNewSources()
         }
     }
 
+    private void addBlockedOperator(List<Operator> blockedOperators, List<ListenableFuture<?>> blockedFutures, Operator operator, ListenableFuture<?> blockedFuture)
+    {
+        if (!blockedOperators.isEmpty() && getLast(blockedOperators) == operator) {
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: explain for in comment why we need this.

File: presto-main/src/main/java/com/facebook/presto/operator/Driver.java (hunk 3 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: Well, when there are a few operators in a driver, some time passes between when we first consulted the future and now, when we construct the `firstFinishedFuture`. `firstFinishedFuture` has visible overhead (it registers a listener in all futures), this is why i wanted to re-check the futures before calling `firstFinishedFuture`. Also, this is more in the spirit of the code before my change: we asked operators anew whether they are blocked.

File: presto-main/src/main/java/com/facebook/presto/operator/Driver.java (hunk 4 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: will do

**Comment by findepi** [comment]
@losipiuk this is not yet ideal. See https://github.com/prestodb/presto/issues/10003#issuecomment-367943174
Let's discuss how to approach this.

File: presto-main/src/main/java/com/facebook/presto/operator/Driver.java (hunk 3 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: Ok. I am fine with that. Just do not see this necessary.

**Comment by losipiuk** [comment]
See travis failure

**Comment by dain** [comment]
Maybe the problem is we need to merge the concepts of `needsInput` and `isBlocked`... so a operator can be blocked waiting for an input page.

**Pull request closed without merging**